cli/replicaset/cmd/upgrade.go

```diff
@@ -0,0 +1,242 @@
+				"and UNIX socket [%s]: %w", fullInstanceName, instance.URI, err)
+		}
+	}
+	return conn, nil
+}
+
+func СollectRWROInfo(replicaset replicaset.Replicaset, master **InstanceMeta,
+	replicas *[]InstanceMeta) error {
+	for _, instance := range replicaset.Instances {
+		run := instance.InstanceCtx
+		fullInstanceName := running.GetAppInstanceName(run)
+		conn, err := getInstanceConnector(instance)
+
+		if err != nil {
+			return err
+		}
+
+		var isRW bool
+		if instance.Mode.String() != "unknown" {
+			isRW = instance.Mode.String() == "rw"
+		} else {
+			res, err := conn.Eval(
+				"return (type(box.cfg) == 'function') or box.info.ro",
+				[]any{}, connector.RequestOpts{})
+			if err != nil {
+				return fmt.Errorf("[%s]: %w", fullInstanceName, err)
+			}
+			isRW = !res[0].(bool)
+		}
+
+		if isRW && *master != nil {
+			return fmt.Errorf("%s and %s are both masters",
+				running.GetAppInstanceName((*master).run), fullInstanceName)
+		} else if isRW {
+			*master = &InstanceMeta{run, conn}
+		} else {
+			*replicas = append(*replicas, InstanceMeta{run, conn})
+		}
+	}
+	return nil
+}
+
+func WaitLSN(conn connector.Connector, masterIID uint32, masterLSN uint64) error {
+	var lastError error
+	query := fmt.Sprintf("return box.info.vclock[%d]", masterIID)
+
+	deadline := time.Now().Add(time.Duration(LsnTimeout) * time.Second)
+	for {
+		res, err := conn.Eval(query, []any{}, connector.RequestOpts{})
+		if err != nil {
+			lastError = fmt.Errorf("failed to evaluate LSN query: %w", err)
+		} else if len(res) == 0 {
+			lastError = errors.New("empty result from LSN query")
+		} else {
+			var lsn uint64
+			if err := mapstructure.Decode(res[0], &lsn); err != nil {
+				lastError = fmt.Errorf("failed to decode LSN: %w", err)
+			} else if lsn >= masterLSN {
+				return nil
+			} else {
+				lastError = fmt.Errorf("current LSN %d is behind required "+
+					"master LSN %d", lsn, masterLSN)
+			}
+		}
+
+		if time.Now().After(deadline) {
+			break
+		}
+
+		time.Sleep(1 * time.Second)
+	}
+
+	return lastError
+}
+
+func upgradeReplicaset(replicaset replicaset.Replicaset) error {
+	var master *InstanceMeta = nil
+	replicas := []InstanceMeta{}
+
+	err := СollectRWROInfo(replicaset, &master, &replicas)
+	if err != nil {
+		return err
+	}
+
+	// upgrade master instance, collect LSN and IID from master instance
+	var upgradeInfo SyncInfo
+	fullMasterName := running.GetAppInstanceName(master.run)
+	res, err := master.conn.Eval(upgradeMasterLua, []any{}, connector.RequestOpts{})
+	if err != nil {
+		return fmt.Errorf("failed to execute upgrade script on master instance - %s: %w",
+			fullMasterName, err)
+	}
+
+	if err := mapstructure.Decode(res[0], &upgradeInfo); err != nil {
+		return fmt.Errorf("failed to decode response from master instance - %s: %w",
+			fullMasterName, err)
+	}
+
+	if upgradeInfo.Err != nil {
+		return fmt.Errorf("master instance upgrade failed - %s: %w",
+			fullMasterName, err)
+	}
```

Comment:
Nit: I like how the `WaitLSN` function wraps the eval request: all decoding and interpretation of the result is inside. We can do the similar `upgradeMaster`, `snapshot` functions that accept a connection and incapsulates the logic of result processing.

This way this block would be like the following:

```go
res, err := upgradeMaster(master.conn)
if err != nil {
	return err
}
```

Maybe we should use some naming pattern for functions that perform actions on an instance? Say, `execute<...>`? `executeUpgrade` and `executeSnapshot` look good, but `waitLSN` doesn't fit there. And maybe it is OK to leave it just `waitLSN`.

Reply:
I chose to name the wrappers `upgradeMaster` and `Snapshot`.
Done.